### ISSUE TITLE
Custom callback functions for notifications in HealthChecks.UI

### ISSUE THREAD

**Issue by Lagostra** [Xabaril/AspNetCore.Diagnostics.HealthChecks#1114]
**What would you like to be added**:
The ability to register C# functions as notification hooks in addition to the existing webhooks, with full access to the name of the relevant liveness and the UIHealthReport. Preferably the callback would also be provided access to the DI container, so that it can access other registered services – e.g. `HealthChecksDb` in order to retrieve information about previous runs and failures for the relevant liveness.

There are many possible ways of implementing this, but a suggestion is allowing the user to provide a `Func<IServiceProvider provider, string livenessName, UIHealthReport report>` to a method on `Settings`:

```C#
builder.Services.AddHealthChecksUI(settings => {
    # Set up health checks and other config

    settings.AddCustomNotification("notification1", (provider, name, report) => {
        # Do custom notification handling here
    });
});
```

Since this callback gives full control over the entire notification process, there is no need for additional methods such as `shouldNotifyFunc` and `customMessageFunc`, as provided on `AddWebhookNotification`.

**Why is this needed**:
We are implementing the HealthCheckUI as a means of monitoring multiple applications and sending notifications upon failures, and have a need for both more advanced filtering of notifications and richer formatting than the current webhook implementation easily allows. Providing a way to run some custom code upon the creation of a notification would allow us to easily implement all these features without requiring custom functionality from the library.

This feature would also facilitate solutions to multiple previously reported issues, such as various notification filtering (#404, #752), and sending emails based on notifications (#560) without requiring calling a controller action in the same solution as the HealthCheckUI.

**Comment by JGud007** [comment]
I agree, this would be very useful to me as well. I would like to be able filter out certain livenessNames from sending notifications and it doesn't seem to be possible with the current _shouldNotifyFunc_ and _customMessageFunc_ as there is no way to check the liveness name via the UIHealthReport class

**Comment by Lagostra** [comment]
@sungam3r I suggest re-opening this issue, since #1153, while definitively useful and a nice low-hanging fruit, does not really solve the use case described in this issue ☺️
